### ISSUE TITLE
Stable `eigh` for degenerate eigenvalues

### ISSUE THREAD

**Issue by jackbaker1001** [XanaduAI/GradDFT#35]
When implementing self-consistent training methods, I thought I had fixed the NaN-gradients problem, but in reality, I only partly fixed it.

The core issue is that the reverse-mode Jacobian is undefined for degenerate eigenproblems. This is because in the eigenvector term for the gradients we have a term like: $1/(\lambda_i - \lambda_j)$ for eigenvalues $\lambda$ which is`+inf` for $\lambda_i = \lambda_j$.

I have implemented a "safe version" of the reverse-mode Jacobian which uses the Lorentzian broadening approach to mitigate this. I used the implementation suggested [here](https://www.nature.com/articles/s42005-021-00568-6), slightly modified for symmetric matrix inputs.

Broadening also only happens when eigenvalue differences are below some tolerance which is set by the user. Therefore, this method produces the same gradients as `jnp.linalg.eigh` for non-degenerate problems.
